File: IPL 2021 - Match Day 2 - GFG/ipl-2021-match-day-2.java
//{ Driver Code Starts
import java.util.*;
import java.io.*;
import java.lang.*;

class Main {
    public static void main(String args[]) {
        // taking input using class Scanner
        Scanner sc = new Scanner(System.in);

        // taking total count of testcases
        int t = sc.nextInt();

        while (t-- > 0) {
            // taking total number of elements
            int n = sc.nextInt();

            // taking size of subArrays
            int k = sc.nextInt();

            // Declaring and Intializing an array of size n
            int arr[] = new int[n];

            // adding all the elements to the array
            for (int i = 0; i < n; i++) {
                arr[i] = sc.nextInt();
            }

            // Calling the method max_of_subarrays of class solve
            // and storing the result in an ArrayList
            ArrayList<Integer> res = new Solution().max_of_subarrays(arr, n, k);

            // printing the elements of the ArrayList
            for (int i = 0; i < res.size(); i++)
                System.out.print(res.get(i) + " ");
            System.out.println();
        }
    }
}
// } Driver Code Ends


class Solution {
    static ArrayList<Integer> max_of_subarrays(int nums[], int n, int k) {
        ArrayList<Integer> result = new ArrayList<>();
        Deque<Integer> dq = new ArrayDeque<>();
        int j = 0;
        for(int i=0; i<k; i++){
            while(!dq.isEmpty() && dq.peekLast() < nums[i]){
                dq.pollLast();
            }
            
            dq.addLast(nums[i]);
        }
        result.add(dq.peek());
        for(int i=k; i<n; i++){
            if(dq.peek() == nums[i-k]){
                dq.poll();
            }
            
            while(!dq.isEmpty() && dq.peekLast() < nums[i]){
                dq.pollLast();
            }
            
            dq.addLast(nums[i]);
            
            result.add(dq.peek());
        }
        return result;
    }
}

File: IPL 2021 - Match Day 2 - GFG/ipl-2021-match-day-2.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
class Solution {
  public:
    vector<int> max_of_subarrays(vector<int> nums, int n, int k) {
        deque<int> dq;
        vector<int> result;

        for(int i=0; i<k; i++){
            while(!dq.empty() && dq.back() < nums[i]){
                dq.pop_back();
            }
            
            dq.push_back(nums[i]);
        }
        result.push_back(dq.front());
        for(int i=k; i<n; i++){
            
            if(dq.front() == nums[i-k]){
                dq.pop_front();
            }
            
            while(!dq.empty() && dq.back() < nums[i]){
                dq.pop_back();
            }
            
            dq.push_back(nums[i]);
            
            result.push_back(dq.front());
        }
        
        return result;
    }
};

//{ Driver Code Starts.

int main() {

    int t;
    cin >> t;

    while (t--) {

        int n, k;
        cin >> n >> k;

        vector<int> arr(n);
        for (int i = 0; i < n; i++) cin >> arr[i];
        Solution ob;
        vector<int> res = ob.max_of_subarrays(arr, n, k);
        for (int i = 0; i < res.size(); i++) cout << res[i] << " ";
        cout << endl;
    }

    return 0;
}
// } Driver Code Ends